lizard_auth_server/backends.py

```diff
@@ -0,0 +1,127 @@
+    user_attrs = dict()
+    for i in attr_list:
+        name = mapping.get(i.get("Name"))
+        if name:
+            value = i.get("Value")
+            user_attrs[name] = value
+    return user_attrs
+
+
+class CognitoUser(Cognito):
+    user_class = get_user_model()
+    # Mapping of Cognito User attribute name to Django User attribute name
+    COGNITO_ATTR_MAPPING = getattr(
+        settings,
+        "COGNITO_ATTR_MAPPING",
+        {
+            "email": "email",
+            "given_name": "first_name",
+            "family_name": "last_name",
+        },
+    )
+
+    def get_user_obj(self, username=None, attribute_list=[], metadata={}, attr_map={}):
+        user_attrs = cognito_to_dict(attribute_list, CognitoUser.COGNITO_ATTR_MAPPING)
+        django_fields = [f.name for f in CognitoUser.user_class._meta.get_fields()]
+        extra_attrs = {}
+        for k, v in user_attrs.items():
+            if k not in django_fields:
+                extra_attrs.update({k: user_attrs.pop(k, None)})
+        if getattr(settings, "COGNITO_CREATE_UNKNOWN_USERS", True):
+            user, created = CognitoUser.user_class.objects.update_or_create(
+                username=username, defaults=user_attrs
+            )
+        else:
+            try:
+                user = CognitoUser.user_class.objects.get(username=username)
+                for k, v in iteritems(user_attrs):
+                    setattr(user, k, v)
+                user.save()
+            except CognitoUser.user_class.DoesNotExist:
+                user = None
+        if user:
+            for k, v in extra_attrs.items():
+                setattr(user, k, v)
+        return user
+
+
+class AbstractCognitoBackend(ModelBackend):
+    __metaclass__ = abc.ABCMeta
+
+    UNAUTHORIZED_ERROR_CODE = "NotAuthorizedException"
+
+    USER_NOT_FOUND_ERROR_CODE = "UserNotFoundException"
+
+    COGNITO_USER_CLASS = CognitoUser
+
+    @abc.abstractmethod
+    def authenticate(self, username=None, password=None):
+        """
+        Authenticate a Cognito User
+        :param username: Cognito username
+        :param password: Cognito password
+        :return: returns User instance of AUTH_USER_MODEL or None
+        """
+        cognito_user = CognitoUser(
+            settings.COGNITO_USER_POOL_ID,
+            settings.COGNITO_APP_ID,
+            access_key=getattr(settings, "AWS_ACCESS_KEY_ID", None),
+            secret_key=getattr(settings, "AWS_SECRET_ACCESS_KEY", None),
+            username=username,
+        )
+        try:
+            cognito_user.authenticate(password)
+        except (Boto3Error, ClientError) as e:
+            return self.handle_error_response(e)
+        user = cognito_user.get_user()
+        if user:
+            user.access_token = cognito_user.access_token
+            user.id_token = cognito_user.id_token
+            user.refresh_token = cognito_user.refresh_token
+
+        return user
+
+    def handle_error_response(self, error):
```

Comment:
Hier moeten ook errors die vanuit de lambda trigger komen worden afgevangen (bijv. invalid username/password). 

Je kan wellicht vanuit de console een authenticatie initieren en een verkeerd username,wachtwoord invoeren (die resulteert in een call naar staging SSO), dan zie je de error message.

Reply:
Je krijgt zoiets terug:

     '{"__type":"UserNotFoundException",
       "message":"UserMigration failed with error \'Incorrect username or password\'."}'

Ik heb in ieder geval nog niets anders eruit weten te krijgen dan die "user not found" en "gaat heen, user/pass klopt niet".

- AD gebruiker met verkeerd wachtwoord
- django gebruiker met verkeerd wachtwoord
- niet-bestaande gebruiker

Dus.... volgens mij is onze trigger inmiddels robuust en netjes genoeg om alles af te vangen.